File: requery-android/src/main/java/io/requery/android/sqlitex/SqlitexDatabaseSource.java
/*
 * Copyright 2017 requery.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.requery.android.sqlitex;

import android.content.Context;
import android.database.Cursor;
import android.text.TextUtils;
import io.requery.android.DefaultMapping;
import io.requery.android.LoggingListener;
import io.requery.android.database.sqlite.SQLiteDatabase;
import io.requery.android.database.sqlite.SQLiteOpenHelper;
import io.requery.android.sqlite.DatabaseProvider;
import io.requery.android.sqlite.SchemaUpdater;
import io.requery.meta.EntityModel;
import io.requery.sql.ConfigurationBuilder;
import io.requery.sql.Mapping;
import io.requery.sql.Platform;
import io.requery.sql.SchemaModifier;
import io.requery.sql.TableCreationMode;
import io.requery.sql.platform.SQLite;
import io.requery.util.function.Function;

import java.sql.Connection;
import java.sql.SQLException;

/**
 * Wrapper for working with Android SQLite databases. That use the
 * <a href="https://github.com/requery/sqlite-android">requery Android SQLite support library</a>.
 *
 * <p> This class extends the support version of {@link SQLiteOpenHelper} and is used to create the
 * database. {@link #onCreate(SQLiteDatabase)} will create the database tables & columns
 * automatically.
 *
 * <p> {@link #onUpgrade(SQLiteDatabase, int, int)} will create any missing tables/columns, however
 * it will not remove any tables/columns more complex upgrades should be handled by overriding
 * {@link #onUpgrade(SQLiteDatabase, int, int)} and implementing a script to handle the migration.
 *
 * @author Nikhil Purushe
 */
public class SqlitexDatabaseSource extends SQLiteOpenHelper implements
    DatabaseProvider<SQLiteDatabase> {

    private final Platform platform;
    private final EntityModel model;
    private Mapping mapping;
    private SQLiteDatabase db;
    private io.requery.sql.Configuration configuration;
    private boolean loggingEnabled;
    private TableCreationMode mode;

    public SqlitexDatabaseSource(Context context, EntityModel model, int version) {
        this(context, model, getDefaultDatabaseName(context, model), version);
    }

    public SqlitexDatabaseSource(Context context,
                                 EntityModel model,
                                 String name,
                                 int version) {
        super(context, name, null, version);
        if (model == null) {
            throw new IllegalArgumentException("null model");
        }
        this.platform = new SQLite();
        this.model = model;
        this.mode = TableCreationMode.CREATE_NOT_EXISTS;
    }

    private static String getDefaultDatabaseName(Context context, EntityModel model) {
        return TextUtils.isEmpty(model.getName()) ?
            context.getPackageName() : model.getName();
    }

    @Override
    public void setLoggingEnabled(boolean enable) {
        this.loggingEnabled = enable;
    }

    @Override
    public void setTableCreationMode(TableCreationMode mode) {
        this.mode = mode;
    }

    protected Mapping onCreateMapping(Platform platform) {
        return new DefaultMapping(platform);
    }

    protected void onConfigure(ConfigurationBuilder builder) {
        if (loggingEnabled) {
            LoggingListener loggingListener = new LoggingListener();
            builder.addStatementListener(loggingListener);
        }
    }

    private Connection getConnection(SQLiteDatabase db) throws SQLException {
        synchronized (this) {
            return new SqlitexConnection(db);
        }
    }

    @Override
    public io.requery.sql.Configuration getConfiguration() {
        if (mapping == null) {
            mapping = onCreateMapping(platform);
        }
        if (mapping == null) {
            throw new IllegalStateException();
        }
        if (configuration == null) {
            ConfigurationBuilder builder = new ConfigurationBuilder(this, model)
                .setMapping(mapping)
                .setPlatform(platform)
                .setBatchUpdateSize(1000);
            onConfigure(builder);
            configuration = builder.build();
        }
        return configuration;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        this.db = db;
        new SchemaModifier(getConfiguration()).createTables(TableCreationMode.CREATE);
    }

    @Override
    public void onConfigure(SQLiteDatabase db) {
        db.setForeignKeyConstraintsEnabled(true);
    }

    @Override
    public void onUpgrade(final SQLiteDatabase db, int oldVersion, int newVersion) {
        this.db = db;
        SchemaUpdater updater = new SchemaUpdater(getConfiguration(),
            new Function<String, Cursor>() {
            @Override
            public Cursor apply(String s) {
                return db.rawQuery(s, null);
            }
        }, mode);
        updater.update();
    }

    @Override
    public Connection getConnection() throws SQLException {
        synchronized (this) {
            if (db == null) {
                db = getWritableDatabase();
            }
            return getConnection(db);
        }
    }
}

File: requery-android/src/main/java/io/requery/android/sqlitex/SqlitexDatabaseSource.kt
/*
 * Copyright 2018 requery.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.requery.android.sqlitex

import android.content.Context
import android.text.TextUtils
import io.requery.android.DefaultMapping
import io.requery.android.LoggingListener
import io.requery.android.database.sqlite.SQLiteDatabase
import io.requery.android.database.sqlite.SQLiteOpenHelper
import io.requery.android.sqlite.DatabaseProvider
import io.requery.android.sqlite.SchemaUpdater
import io.requery.meta.EntityModel
import io.requery.sql.*
import io.requery.sql.platform.SQLite
import io.requery.util.function.Function

import java.sql.Connection
import java.sql.SQLException

/**
 * Wrapper for working with Android SQLite databases. That use the
 * [requery Android SQLite support library](https://github.com/requery/sqlite-android).
 *
 *
 *  This class extends the support version of [SQLiteOpenHelper] and is used to create the
 * database. [.onCreate] will create the database tables & columns
 * automatically.
 *
 *
 *  [.onUpgrade] will create any missing tables/columns, however
 * it will not remove any tables/columns more complex upgrades should be handled by overriding
 * [.onUpgrade] and implementing a script to handle the migration.
 *
 * @author Nikhil Purushe
 */
open class SqlitexDatabaseSource(context: Context,
                            private val model: EntityModel?,
                            name: String,
                            version: Int)
    : SQLiteOpenHelper(context, name, null, version), DatabaseProvider<SQLiteDatabase> {

    private val platform: Platform
    private var mapping: Mapping? = null
    private var db: SQLiteDatabase? = null
    private var _configuration: io.requery.sql.Configuration? = null
    private var loggingEnabled: Boolean = false
    private var mode: TableCreationMode? = null

    constructor(context: Context, model: EntityModel, version: Int)
            : this(context, model, getDefaultDatabaseName(context, model), version)

    init {
        if (model == null) {
            throw IllegalArgumentException("null model")
        }
        this.platform = SQLite()
        this.mode = TableCreationMode.CREATE_NOT_EXISTS
    }

    companion object {
        private fun getDefaultDatabaseName(context: Context, model: EntityModel): String {
            return if (TextUtils.isEmpty(model.name))
                context.packageName
            else
                model.name
        }
    }

    override fun setLoggingEnabled(enable: Boolean) {
        this.loggingEnabled = enable
    }

    override fun setTableCreationMode(mode: TableCreationMode) {
        this.mode = mode
    }

    protected fun onCreateMapping(platform: Platform): Mapping {
        return DefaultMapping(platform)
    }

    protected fun onConfigure(builder: ConfigurationBuilder) {
        if (loggingEnabled) {
            val loggingListener = LoggingListener()
            builder.addStatementListener(loggingListener)
        }
    }

    @Throws(SQLException::class)
    private fun getConnection(db: SQLiteDatabase): Connection {
        synchronized(this) {
            return SqlitexConnection(db)
        }
    }

    override fun getConfiguration(): Configuration {
        if (mapping == null) {
            mapping = onCreateMapping(platform)
        }
        if (mapping == null) {
            throw IllegalStateException()
        }
        if (_configuration == null) {
            val builder = ConfigurationBuilder(this, model)
                    .setMapping(mapping)
                    .setPlatform(platform)
                    .setBatchUpdateSize(1000)
            onConfigure(builder)
            _configuration = builder.build()
        }
        return _configuration!!
    }

    override fun onCreate(db: SQLiteDatabase) {
        this.db = db
        SchemaModifier(configuration).createTables(TableCreationMode.CREATE)
    }

    override fun onConfigure(db: SQLiteDatabase) {
        db.setForeignKeyConstraintsEnabled(true)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        this.db = db
        val updater = SchemaUpdater(configuration,
                Function { s -> db.rawQuery(s, null) }, mode)
        updater.update()
    }

    @Throws(SQLException::class)
    override fun getConnection(): Connection {
        synchronized(this) {
            if (db == null) {
                db = getWritableDatabase()
            }
            return getConnection(db!!)
        }
    }
}
